### PR TITLE
Regression tests for the Distributed module 

### PR DESCRIPTION
**Pull request by simone-silvestri**
This PR adapts the hydrostatic regression tests to be run in distributed mode.

an important step in this direction is to allow array partitioning under the hood in `set!` if the array size coincides with the global size of the field (implemented here)

a more important step is to remove the topology from the `Distributed` architecture (it should not know anything about topology which is a grid property)

### CODE REVIEW

File: src/DistributedComputations/distributed_grids.jl [hunk 1]
```diff
@@ -51,10 +54,10 @@ function RectilinearGrid(arch::Distributed,
                          extent = nothing,
                          topology = (Periodic, Periodic, Bounded))
 
-    TX, TY, TZ, global_size, halo, x, y, z =
+    TX, TY, TZ, gsize, halo, x, y, z =
```
Comment (navidcy) [inline src/DistributedComputations/distributed_grids.jl:57]: `global_size` is more verbose, why change?

Comment (simone-silvestri) [reply src/DistributedComputations/distributed_grids.jl:57]: don't want to conflate it with the function `global_size`

Comment (navidcy) [reply src/DistributedComputations/distributed_grids.jl:57]: gotcha!

File: src/DistributedComputations/distributed_grids.jl [hunk 2]
```diff
@@ -19,15 +19,15 @@ const DistributedRectilinearGrid{FT, TX, TY, TZ, FX, FY, FZ, VX, VY, VZ} =
...
+function local_size(p::Partition, topo, ranks, gsize)
     # TODO: check correctness
     return p.sizes[rank]
 end
 
 const EqualPartition = Partition{Nothing}
 
-function local_size(p::EqualPartition, rank, global_size)
-    Nx, Ny, Nz = global_size
+function local_size(p::EqualPartition, rank, gsize)
```
Comment (glwagner) [inline src/DistributedComputations/distributed_grids.jl:29]:
```suggestion
function local_size(p::EqualPartition, rank, global_size)
```

**Comment by glwagner** [comment]
Can you change `gsize` to `global_size` everywhere?

**Comment by glwagner** [comment]
Can you explain how topology was removed? I was under the impression this was necessary for determining connectivity (or something like that) but clearly this is wrong. A detailed explanation would be helpful...

File: src/DistributedComputations/distributed_grids.jl (hunk 2 above)
Comment (navidcy) [reply src/DistributedComputations/distributed_grids.jl:29]: https://github.com/CliMA/Oceananigans.jl/pull/3328#discussion_r1355163735

**Comment by simone-silvestri** [comment]
The connectivity creates a "cartesian grid" of ranks. 
This type informs us on which ranks are next to each other. 
The more general way to arrange this grid, in absence of other information is to wrap around the ranks as if the cartesian rank grid was `Periodic`.

This is enough information for parsing the rank configuration, but not enough to perform fill halo regions. 
For this, we need a grid! Only with a grid, we will know if some directions are Bounded or Periodic, so, leveraging this information we can correctly assess which boundaries need communicating or not.

Philosophically speaking, architecture provides the general rank layout while the grid concretizes this information 
leveraging the topology to decide whether to perform halo passing or not

Practically speaking the implementation leverages an `inject_halo_communication_boundary_conditions` implemented here:
https://github.com/CliMA/Oceananigans.jl/blob/34a3b930e0ace7df7dc7660f9c74f52315f41da7/src/DistributedComputations/halo_communication_bcs.jl#L14
previously only the connectivity was passed to this function, while also logically also grid information is required.

**Comment by glwagner** [comment]
> The connectivity creates a "cartesian grid" of ranks which informs us which ranks are spatial neighbours.

Okay, but now this is only partially true, because the connectivity cannot know that two ranks are neighbors across a Periodic boundary.

I'm totally fine with this concept of connectivity, but just wanted to clarify that. Basically there are additional connections implied by Periodic which cannot be resolved by the architecture alone. But since those connections don't have to be resolved until after the grid is created (they are resolved within `inject_halo_communication_bcs`), it's not necessary to store the "total" connectivity inside the architecture.

**Comment by simone-silvestri** [comment]
Should be ready to merge when tests pass

File: src/DistributedComputations/partition_assemble.jl [hunk 3]
```diff
@@ -5,6 +5,10 @@ all_reduce(val, arch::Distributed; op = +) =
 
 all_reduce(val, arch; kwargs...) = val
 
+# MPI Barrier
+barrier!(arch) = nothing
+barrier!(arch::Distributed) = MPI.Barrier(arch.communicator)
```
Comment (glwagner) [inline src/DistributedComputations/partition_assemble.jl:10]: Should this be just `barrier`? not `barrier!`.

File: src/DistributedComputations/distributed_fields.jl [hunk 4]
```diff
@@ -33,6 +35,23 @@ function set!(u::DistributedField, f::Function)
     return u
 end
 
+# Automatically partition under the hood if sizes are compatible
+function set!(u::DistributedField, v::Union{Array, CuArray})
+    gsize = global_size(architecture(u), size(u))
```
Comment (glwagner) [inline src/DistributedComputations/distributed_fields.jl:40]:
```suggestion
    global_sz = global_size(architecture(u), size(u))
```

File: src/DistributedComputations/distributed_grids.jl (hunk 1 above)
Comment (glwagner) [reply src/DistributedComputations/distributed_grids.jl:57]: use `global_sz` consistent with how we use `sz` elsewhere

File: test/regression_tests/hydrostatic_free_turbulence_regression_test.jl [hunk 5]
```diff
@@ -1,10 +1,26 @@
...
 using Oceananigans.TurbulenceClosures: HorizontalScalarDiffusivity
 
+using Oceananigans.DistributedComputations: DistributedGrid, DistributedComputations, all_reduce
+using Oceananigans.DistributedComputations: reconstruct_global_topology, partition_global_array, cpu_architecture
+
 using JLD2
 
+ordered_indices(r, i) = i == 1 ? r : i == 2 ? (r[2], r[1], r[3]) : (r[3], r[2], r[1])
+
+get_topology(grid, i) = string(topology(grid, i))
```
Comment (glwagner) [inline test/regression_tests/hydrostatic_free_turbulence_regression_test.jl:14]:
```suggestion
global_topology(grid, i) = string(topology(grid, i))
```

File: src/DistributedComputations/halo_communication.jl [hunk 6]
```diff
@@ -116,13 +115,13 @@ end
...
+    # if `isnothing(requests)`, `fill_halo!` did not involve MPI passing
     if isnothing(requests)
         return nothing
     end
 
     # Overlapping communication and computation, store requests in a `MPI.Request`
-    # pool to be waited upon after tendency calculation
+    # pool to be waited upon later on when halos are required.
     if async && !(arch isa SynchronizedDistributed)
         push!(arch.mpi_requests, requests...)
```
Comment (glwagner) [inline src/DistributedComputations/halo_communication.jl:126]:
```suggestion
        append!(arch.mpi_requests, requests)
```

File: test/regression_tests/hydrostatic_free_turbulence_regression_test.jl [hunk 7]
```diff
@@ -33,6 +49,7 @@ function run_hydrostatic_free_turbulence_regression_test(grid, free_surface; reg
 
     u, v, w = model.velocities
     U       = 0.1 * maximum(abs, u)
+    U       = all_reduce(U, architecture(grid), op = max)
```
Comment (glwagner) [inline test/regression_tests/hydrostatic_free_turbulence_regression_test.jl:52]: xref https://github.com/CliMA/Oceananigans.jl/issues/3337

File: src/DistributedComputations/partition_assemble.jl [hunk 8]
```diff
@@ -5,6 +5,10 @@ all_reduce(val, arch::Distributed; op = +) =
 
 all_reduce(val, arch; kwargs...) = val
```
Comment (glwagner) [inline src/DistributedComputations/partition_assemble.jl:6]: I suggest

```julia
all_reduce(op, A, arch)
```

to mirror the syntax for `reduce`.

File: src/DistributedComputations/halo_communication_bcs.jl [hunk 9]
```diff
@@ -27,10 +27,21 @@ function inject_halo_communication_boundary_conditions(field_bcs, local_rank, co
...
+    # `LeftConnected` means bounded on the right
+    inject_west  = !isnothing(rank_west)  && (TX != RightConnected) 
+    inject_east  = !isnothing(rank_east)  && (TX != LeftConnected) 
+    inject_south = !isnothing(rank_south) && (TY != RightConnected) 
+    inject_north = !isnothing(rank_north) && (TY != LeftConnected) 
+
+    west     = inject_west  ? west_comm_bc  : field_bcs.west  
+    east     = inject_east  ? east_comm_bc  : field_bcs.east  
+    south    = inject_south ? south_comm_bc : field_bcs.south 
+    north    = inject_north ? north_comm_bc : field_bcs.north
```
Comment (navidcy) [inline src/DistributedComputations/halo_communication_bcs.jl:44]:
```suggestion
    west  = inject_west  ? west_comm_bc  : field_bcs.west  
    east  = inject_east  ? east_comm_bc  : field_bcs.east  
    south = inject_south ? south_comm_bc : field_bcs.south 
    north = inject_north ? north_comm_bc : field_bcs.north 
```

File: test/test_distributed_hydrostatic_model.jl [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+        cpu_arch = cpu_architecture(arch)
+
+        @test all(isapprox(u, partition_global_array(cpu_arch, us, size(u)), atol=1e-20, rtol = 1e-15))
+        @test all(isapprox(v, partition_global_array(cpu_arch, vs, size(v)), atol=1e-20, rtol = 1e-15))
+        @test all(isapprox(w, partition_global_array(cpu_arch, ws, size(w)), atol=1e-20, rtol = 1e-15))
+        @test all(isapprox(c, partition_global_array(cpu_arch, cs, size(c)), atol=1e-20, rtol = 1e-15))
+        @test all(isapprox(η, partition_global_array(cpu_arch, ηs, size(η)), atol=1e-20, rtol = 1e-15))
+    end
+end
```
Comment (navidcy) [inline test/test_distributed_hydrostatic_model.jl:115]:
```suggestion
end

```

File: test/test_distributed_hydrostatic_model.jl [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+                                        halo = (3, 3, 3),
+                                        latitude = (-80, 80),
+                                        longitude = (-160, 160),
+                                        z = (-1, 0),
+                                        radius = 1,
+                                        topology=(Bounded, Bounded, Bounded))
+
+        global_grid = reconstruct_global_grid(grid)
+
+        us, vs, ws, cs, ηs = solid_body_rotation_test(global_grid)
```
Comment (glwagner) [inline test/test_distributed_hydrostatic_model.jl:90]:
```suggestion
        # "s" for "serial" computation
        us, vs, ws, cs, ηs = solid_body_rotation_test(global_grid)
```

File: test/test_distributed_hydrostatic_model.jl [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        v = Array(interior(v))
+        w = Array(interior(w))
+        c = Array(interior(c))
+        η = Array(interior(η))
+
+        @test all(isapprox(u, partition_global_array(arch, us, size(u)), atol=1e-20, rtol = 1e-15))
+        @test all(isapprox(v, partition_global_array(arch, vs, size(v)), atol=1e-20, rtol = 1e-15))
+        @test all(isapprox(w, partition_global_array(arch, ws, size(w)), atol=1e-20, rtol = 1e-15))
+        @test all(isapprox(c, partition_global_array(arch, cs, size(c)), atol=1e-20, rtol = 1e-15))
+        @test all(isapprox(η, partition_global_array(arch, ηs, size(η)), atol=1e-20, rtol = 1e-15))
```
Comment (glwagner) [inline test/test_distributed_hydrostatic_model.jl]: maybe call `partition_global_array(arch, us, size(u))` on a different line to make this easier to read

File: test/utils_for_runtests.jl [hunk 13]
```diff
@@ -6,10 +6,25 @@ using Test
...
+function test_architectures() 
+    child_arch =  CUDA.has_cuda() ? GPU() : CPU()
+
+    # If MPI is initialized with MPI.Comm_size > 0, we are running in parallel.
+    # We test 3 different configurations: `Partition(x = 4)`, `Partition(y = 4)` 
+    # and `Partition(x = 4, y = 4)`
+    if MPI.Initialized() && MPI.Comm_size(MPI.COMM_WORLD) == 4
+        return tuple(Distributed(child_arch; partition = Partition(4)))
+                # Distributed(child_arch; partition = Partition(1, 2)))
+                #Distributed(child_arch; partition = Partition(2, 2)))
```
Comment (glwagner) [inline test/utils_for_runtests.jl:23]:
```suggestion
                # TODO: support y-partitions and xy-partitions
                # Distributed(child_arch; partition = Partition(1, 2)))
                # Distributed(child_arch; partition = Partition(2, 2)))
```

**Review by glwagner — APPROVED** [review]

File: test/regression_tests/hydrostatic_free_turbulence_regression_test.jl [hunk 14]
```diff
@@ -47,10 +64,11 @@ function run_hydrostatic_free_turbulence_regression_test(grid, free_surface; reg
     CUDA.allowscalar(false)
 
     wave_time_scale = min(minimum_Δx, minimum_Δy) / wave_speed
-
+
```
Comment (navidcy) [inline test/regression_tests/hydrostatic_free_turbulence_regression_test.jl:67]:
```suggestion

```

**Comment by glwagner** [comment]
merge when ready

**Pull request merged**